src/translate.py

```diff
@@ -431,29 +448,89 @@ def get_stack_info(
+    for node in flow_graph.nodes:
+        for inst in node.block.instructions:
+            if not inst.args or not isinstance(inst.args[0], Register):
+                continue
+            destination = inst.args[0]
 
-    # Find the region that contains local variables. It is above saved registers
-    # and the return address, if those exist. If they don't, the local variables
-    # can lie directly at the bottom of the stack.
-    info.local_vars_region_bottom = 0
+            if (
+                inst.mnemonic in ["lw", "lwc1", "ldc1"]
+                and isinstance(inst.args[1], AsmAddressMode)
+                and inst.args[1].rhs.register_name == "sp"
+                and inst.args[1].lhs_as_literal() > 16
+            ):
+                info.subroutine_arg_top = update_min(
+                    info.subroutine_arg_top, inst.args[1].lhs_as_literal()
+                )
+            elif (
+                inst.mnemonic == "addiu"
+                and destination.register_name != "sp"
+                and inst.args[1] == Register("sp")
+                and isinstance(inst.args[2], AsmLiteral)
+                and inst.args[2].value < info.allocated_stack_size
+            ):
+                info.subroutine_arg_top = update_min(
+                    info.subroutine_arg_top, inst.args[2].value
+                )
 
     if not info.is_leaf:
-        info.local_vars_region_bottom = info.return_addr_location + 4
+        # Compute the bounds of the callee-saved register region, including padding
+        callee_saved_offset_and_size.sort()
+        bottom, last_size = callee_saved_offset_and_size[0]
+
+        # Both IDO & GCC save registers in two subregions:
+        # (a) One for double-sized registers
+        # (b) One for word-sized registers, padded to a multiple of 8 bytes
+        # IDO has (a) lower than (b); GCC has (b) lower than (a)
+        # Check that there are no gaps in this region, other than a single
+        # 4-byte word between subregions.
+        top = bottom
+        internal_padding_added = False
+        for offset, size in callee_saved_offset_and_size:
+            if offset != top:
+                if (
+                    not internal_padding_added
+                    and size != last_size
+                    and offset == top + 4
+                ):
+                    top += 4
+                    internal_padding_added = True
+                else:
+                    raise DecompFailure(
+                        f"Gap in callee-saved word stack region. "
+                        f"Saved: {callee_saved_offset_and_size}, "
+                        f"gap at: {offset} != {top}."
+                    )
+            top += size
+            last_size = size
+        # Expand boundaries to multiples of 8 bytes
+        bottom -= 4 if bottom % 8 == 4 else 0
+        top += 4 if top % 8 == 4 else 0
+        info.callee_save_reg_region = (bottom, top)
```

Comment:
I know this whole bit is a lot stricter than it needs to be to function -- but the code is making a lot of assumptions/guesses, so I'd rather quickly identify corner cases that break these assumptions.